geochem_inverse_optimize.py

```diff
@@ -281,6 +280,279 @@ def get_roughness(self) -> float:
+
+        # Solvers that can handle this problem type include:
+        # ECOS, SCS
+        # See: https://www.cvxpy.org/tutorial/advanced/index.html#choosing-a-solver
+        # See: https://www.cvxpy.org/tutorial/advanced/index.html#setting-solver-options
+        solvers = {
+            # VERY SLOW, probably don't use
+            "scip": {
+                "solver": cp.SCIP,
+                "verbose": True,
+            },
+            "ecos": {
+                "solver": cp.ECOS,
+                "verbose": True,
+                "max_iters": 10000,
+                "abstol_inacc": 5e-5,
+                "reltol_inacc": 5e-5,
+                "feastol_inacc": 1e-4,
+            },
+            "scs": {"solver": cp.SCS, "verbose": True, "max_iters": 10000},
+            "gurobi": {"solver": cp.GUROBI, "verbose": False, "NumericFocus": 3},
+        }
+        objective_value = self._problem.solve(**solvers[solver])
+        print(
+            "{color}Status = {status}\033[39m".format(
+                color="" if self._problem.status == "optimal" else "\033[91m",
+                status=self._problem.status,
+            )
+        )
+        print(f"Objective value = {objective_value}")
+        # Return outputs
+        downstream_preds = self.get_downstream_prediction_dictionary()
+        downstream_preds = {sample: value * obs_mean for sample, value in downstream_preds.items()}
+        upstream_preds = self.get_upstream_prediction_map() * obs_mean
+        return downstream_preds, upstream_preds
+
+    def get_downstream_prediction_dictionary(self) -> ElementData:
+        # Print the solution we found
+        predictions: ElementData = {}
+        for sample_name, data in self.sample_network.nodes(data=True):
+            data = data["data"]
+            predictions[sample_name] = data.total_flux.value / data.total_upstream_area
+        return predictions
+
+    def get_upstream_prediction_map(self) -> np.array:
+        out = np.zeros(self.grid.area_labels.shape)
+        xstep = out.shape[1] / self.grid.nx
+        ystep = out.shape[0] / self.grid.ny
+        out = np.zeros(self.grid.area_labels.shape)
+        # Loop through inversion grid nodes
+        for i in np.arange(self.grid.nx):
+            for j in np.arange(self.grid.ny):
+                # indices which subdivide the areas on the base array for each inversion grid.
+                x_start = int(np.floor(i * xstep))
+                x_end = int(np.floor((i + 1) * xstep))
+                y_start = int(np.floor(j * ystep))
+                y_end = int(np.floor((j + 1) * ystep))
+                node = self.grid.node_coord_dict[(self.grid.xs[i], self.grid.ys[j])]
+                # Catch exception for nodes outside of area
+                if node.concentration:
+                    val = node.concentration.value
+                else:
+                    val = np.nan
+                out[y_start:y_end, x_start:x_end] = val
```

Comment:
See comment [here](https://github.com/r-barnes/faster-unmixer/pull/24#discussion_r1035782643). Essentially, each `InverseNode` corresponds to a rectangle of pixels on the base DEM. The `x` and `y` values stored in `InverseNode` correspond to the centre of the rectangle.

Reply:
Making a comment to this effect might be good.

There should also probably be a check to prevent a zero or negative number of pixels from being put in a rectangle if that's possible.